contracts/token/Faucet.sol

```diff
@@ -105,6 +125,76 @@ contract Faucet {
         emit FaucetDripped(_recipient);
     }
 
+    /**
+     * @notice Drip ERC20 tokens to a list of addresses as a batch to decrease gas costs.
+     * There is no check to see if the addresses have already been dripped to. Use with caution.
+     * @dev This function is a modified version of the airdropERC20 function from GasliteDrop.sol https://etherscan.io/address/0x09350f89e2d7b6e96ba730783c2d76137b045fef#code
+     *
+     * @param _token The address of the ERC20 contract
+     * @param _addresses The addresses to airdrop to
+     * @param _amounts The amounts to airdrop
+     * @param _totalAmount The total amount to airdrop
+     */
+    function batchDrip(
+        address _token,
+        address[] calldata _addresses,
+        uint256[] calldata _amounts,
+        uint256 _totalAmount
+    ) external payable isApprovedOperator {
+        emit BatchDrip(_addresses.length, _totalAmount);
+
+        assembly {
+            // Check that the number of addresses matches the number of amounts
+            if iszero(eq(_amounts.length, _addresses.length)) {
+                revert(0, 0)
+            }
+
+            // transferFrom(address from, address to, uint256 amount)
+            mstore(0x00, hex"23b872dd")
+            // from address
+            mstore(0x04, caller())
+            // to address (this contract)
+            mstore(0x24, address())
+            // total amount
+            mstore(0x44, _totalAmount)
+
+            // transfer total amount to this contract
+            if iszero(call(gas(), _token, 0, 0x00, 0x64, 0, 0)) {
+                revert(0, 0)
+            }
+
+            // transfer(address to, uint256 value)
+            mstore(0x00, hex"a9059cbb")
+
+            // end of array
+            let end := add(_addresses.offset, shl(5, _addresses.length))
+            // diff = _addresses.offset - _amounts.offset
+            let diff := sub(_addresses.offset, _amounts.offset)
+
+            // Loop through the addresses
+            for {
+                let addressOffset := _addresses.offset
+            } 1 {
+
+            } {
+                // to address
+                mstore(0x04, calldataload(addressOffset))
+                // amount
+                mstore(0x24, calldataload(sub(addressOffset, diff)))
+                // transfer the tokens
+                if iszero(call(gas(), _token, 0, 0x00, 0x64, 0, 0)) {
+                    revert(0, 0)
```

Comment:
`revert(0,0)` won't return anything readable, you need to point to the memory address of a string for the revert to be interpretable:

https://ethereum.stackexchange.com/questions/133748/trying-to-understand-solidity-assemblys-revert-function

Reply:
you should be able to store the error message outside the assembly block and just reference it in here. so it'd look like.

```
string message = "I'm reverting";

assembly {
...

revert(add(32, message), mload(message))
}
```

Now, will this show up in hardhat as expected? I don't know, but this how the message would be referenced.

Reply:
in both cases here, you're doing a call, though, so you should be able to pull out the revert message from the call instead like this: https://ethereum.stackexchange.com/questions/69133/forward-revert-message-from-low-level-solidity-call

Reply:
since the call itself is in assembly, you have to capture the reason as part of the output data of the call which is encoded in the last two parameters of the function (first pointer of where to write it and then size which is always 0x20)

Reply:
I think you can also do something like:
```
if success {
                returndatacopy(0, 0, returndatasize())
                revert(0, returndatasize())
}
```